apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/helper/ConnectionMetaData.java

```diff
@@ -0,0 +1,434 @@
+     */
+    public static ConnectionMetaData create(String connectionUrl, String user) {
+        String dbVendor = "unknown";
+
+        // trimming a temp copy, keeping the original for logging purposes
+        String tmpUrl = connectionUrl;
+
+        // Connection URLs have a common prefix, starting with "jdbc:", followed by the vendor name and a colon.
+        // The rest is vendor specific.
+        //
+        // Examples:
+        // jdbc:postgresql://hostname/db?user=jdo&password=pass
+        // jdbc:sqlserver://localhost:32958;sslProtocol=TLS;jaasConfigurationName=SQLJDBCDriver
+        // jdbc:oracle:oci:root/secret@localhost:1521:testdb
+        // jdbc:derby:memory:testdb
+        // jdbc:h2:mem:test
+        int indexOfJdbc = tmpUrl.indexOf("jdbc:");
+
+        if (indexOfJdbc != -1) {
+            tmpUrl = tmpUrl.substring(indexOfJdbc + 5);
+            int indexOfNextColon = tmpUrl.indexOf(":");
+            if (indexOfNextColon != -1) {
+                dbVendor = tmpUrl.substring(0, indexOfNextColon);
+                tmpUrl = tmpUrl.substring(indexOfNextColon + 1);
+            }
+        }
+
+        // Further parsing needs to be vendor specific.
+        ConnectionMetaData ret = null;
+        ConnectionUrlParser connectionUrlParser = parsers.get(dbVendor);
+        if (connectionUrlParser != null) {
+            try {
+                ret = connectionUrlParser.parse(tmpUrl, user);
+            } catch (Exception e) {
+                logger.error("Failed to parse connection URL", e);
+            }
+        } else {
+            // Doesn't hurt to try...
+            ret = ConnectionUrlParser.defaultParse(connectionUrl, dbVendor, -1, user);
+        }
+
+        if (ret == null) {
+            ret = new ConnectionMetaData(dbVendor, null, -1, user);
+        }
+        if (logger.isDebugEnabled()) {
+            logger.debug("Based on the connection URL {}, parsed metadata is: {}", connectionUrl, ret);
+        }
+        return ret;
+    }
+
+    private final String dbVendor;
+    @Nullable
+    private final String host;
+    private final int port;
+    private final String user;
+
+    private ConnectionMetaData(String dbVendor, @Nullable String host, int port, String user) {
+        this.dbVendor = dbVendor;
+        this.host = host;
+        this.port = port;
+        this.user = user;
+    }
+
+    public String getDbVendor() {
+        return dbVendor;
+    }
+
+    @Nullable
+    public String getHost() {
+        return host;
+    }
+
+    public int getPort() {
+        return port;
+    }
+
+    public String getUser() {
+        return user;
+    }
+
+    @Override
+    public String toString() {
+        return "ConnectionMetaData{" +
+            "dbVendor='" + dbVendor + '\'' +
+            ", host='" + host + '\'' +
+            ", port=" + port +
+            ", user='" + user + '\'' +
+            '}';
+    }
+
+    @SuppressWarnings("unused")
+    private enum ConnectionUrlParser {
```

Comment:
Oof, I wish there was a standard for those URLs or a library we could use to offload the parsing. @axw are you aware of anything of that sort by any chance?

Reply:
Nope, sorry. For the Go agent I have driver-specific logic which either uses some code from the database driver library to parse, or vendors internal code from the driver library when there's no exported parser function.

Reply:
Imagine how much I wished for such 😅. Couldn't find anything useful. I also thought of using driver-internal APIs, but I figured the URL syntax is expected to be much more stable than such APIs, so went for parsing. I hope it won't bite us too much.